api/wasm/cpp/proxy_wasm_api.h

```diff
@@ -485,24 +500,74 @@ inline bool getStringValue(std::initializer_list<StringView> parts, std::string*
   return true;
 }
 
-inline bool getStructValue(std::initializer_list<StringView> parts,
-                           google::protobuf::Struct* value_ptr) {
+// Specialization for message types (including struct value for lists and maps)
+template <typename T>
+inline bool getMessageValue(std::initializer_list<StringView> parts,
+                            T* value_ptr) {
   auto buf = getProperty(parts);
   if (!buf.has_value()) {
     return false;
   }
+  if (buf.value()->size() == 0) {
+    value_ptr = nullptr;
+    return true;
+  }
   return value_ptr->ParseFromArray(buf.value()->data(), buf.value()->size());
 }
 
-template <typename T> inline bool getValue(std::initializer_list<StringView> parts, T* out) {
-  auto buf = getProperty(parts);
+inline WasmResult exprCreate(StringView expr, uint32_t* token) {
+  return proxy_expr_create(expr.data(), expr.size(), token);
+}
+
+inline Optional<WasmDataPtr> exprEval(uint32_t token) {
+  const char* value_ptr = nullptr;
+  size_t value_size = 0;
+  auto result = proxy_expr_eval(token, &value_ptr, &value_size);
+  if (result != WasmResult::Ok) {
+    return {};
+  }
+  return std::make_unique<WasmData>(value_ptr, value_size);
+}
+
+inline WasmResult exprDelete(uint32_t token) {
+  return proxy_expr_delete(token);
+}
+
+template <typename T>
+inline bool evaluate(uint32_t token, T* out) {
+  auto buf = exprEval(token);
   if (!buf.has_value() || buf.value()->size() != sizeof(T)) {
     return false;
   }
   *out = *reinterpret_cast<const T*>(buf.value()->data());
   return true;
 }
 
+template <>
+inline bool evaluate<std::string>(uint32_t token, std::string* out) {
+  auto buf = exprEval(token);
+  if (!buf.has_value()) {
+    return false;
+  }
+  out->assign(buf.value()->data(), buf.value()->size());
+  return true;
+}
+
+// Specialization for message types (including struct value for lists and maps)
+template <typename T>
+inline bool evaluateMessage(uint32_t token,
+                            T* value_ptr) {
+  auto buf = exprEval(token);
+  if (!buf.has_value()) {
+    return false;
+  }
+  if (buf.value()->size() == 0) {
```

Comment:
code can be de-duplicated btw evaluateMessage and getMessage?
and similar for evaluate and getvalue for string?

Reply:
It sprawled over two headers. We probably need to organize the header better.